Python/Product/TestAdapter/TestExecutor.cs

```diff
@@ -100,293 +142,568 @@ class TestExecutor : ITestExecutor {
             IRunContext runContext,
             IFrameworkHandle frameworkHandle
         ) {
-            // .pyproj file path -> project settings
-            var sourceToSettings = new Dictionary<string, PythonProjectSettings>();
+            bool codeCoverage = EnableCodeCoverage(runContext);
+            string covPath = null;
+            if (codeCoverage) {
+                covPath = GetCoveragePath(tests);
+            }
+            // .py file path -> project settings
+            var sourceToSettings = GetSourceToSettings(runContext.RunSettings);
 
-            foreach (var test in tests) {
+            foreach (var testGroup in tests.GroupBy(x => sourceToSettings[x.CodeFilePath])) {
                 if (_cancelRequested.WaitOne(0)) {
                     break;
                 }
 
-                try {
-                    RunTestCase(frameworkHandle, runContext, test, sourceToSettings);
-                } catch (Exception ex) {
-                    frameworkHandle.SendMessage(TestMessageLevel.Error, ex.ToString());
+                var runner = new TestRunner(
+                    frameworkHandle,
+                    runContext,
+                    tests,
```

Comment:
Yes!


---
In reply to: [67386465](https://github.com/Microsoft/PTVS/pull/1337#discussion_r67386465) [](ancestors = 67386465)